instrumentation/mongo/src/main/java/brave/mongo/TraceMongoCommandListener.java

```diff
@@ -0,0 +1,163 @@
+
+    span.name(getSpanName(commandName, collectionName))
+      .kind(Span.Kind.CLIENT)
+      .remoteServiceName(databaseName)
+      .tag("db.type", "mongo")
+      .tag("mongo.database", databaseName)
+      .tag("mongo.operation", commandName);
+
+    getAbbreviatedCommand(command).ifPresent(abbreviatedCommand -> span.tag("mongo.command", abbreviatedCommand));
+
+    collectionName.ifPresent(collection -> span.tag("mongo.collection", collection));
+
+    final InetSocketAddress socketAddress = event.getConnectionDescription().getServerAddress().getSocketAddress();
+    span.remoteIpAndPort(socketAddress.getAddress().getHostAddress(), socketAddress.getPort());
+
+    span.start();
+  }
+
+  @Override public void commandSucceeded(final CommandSucceededEvent event) {
+    final Span span = activeSpans.remove(event.getRequestId());
+    if (span == null) {
+      return;
+    }
+    span.finish();
+  }
+
+  @Override public void commandFailed(final CommandFailedEvent event) {
+    final Span span = activeSpans.remove(event.getRequestId());
+    if (span == null) {
+      return;
+    }
+    span.error(event.getThrowable());
+    span.finish();
+  }
+
+  static Optional<String> getCollectionName(final BsonDocument command) {
+    final Iterator<Map.Entry<String, BsonValue>> iterator = command.entrySet().iterator();
+    if (!iterator.hasNext()) {
+      return Optional.empty();
+    }
+    final BsonValue value = iterator.next().getValue();
+    if (!value.isString()) {
+      return Optional.empty();
+    }
+    final String stringValue = value.asString().getValue().trim();
+    return stringValue.isEmpty() ? Optional.empty() : Optional.of(stringValue);
+  }
+
+  /**
+   * Returns an abbreviated version of the command for logging/tracing purposes. Currently this is simply a
+   * truncated version of the command's JSON string representation.
+   *
+   * Note that sensitive data related to the Mongo protocol itself is already scrubbed at this point according to
+   * https://github.com/mongodb/specifications/blob/master/source/command-monitoring/command-monitoring.rst#security
+   *
+   * @return an abbreviated version of the command for logging/tracing purposes
+   */
+  Optional<String> getAbbreviatedCommand(final BsonDocument command) {
+    if (maxAbbreviatedCommandLength <= 0) {
+      return Optional.empty();
+    }
+    final String abbreviatedCommand = command.toJson(jsonWriterSettings);
+    return abbreviatedCommand.isEmpty() ? Optional.empty() : Optional.of(abbreviatedCommand);
+  }
+
+  static String getSpanName(final String commandName, final Optional<String> collectionName) {
+    return collectionName.map(collection -> commandName + " " + collection).orElse(commandName);
+  }
+
+  public static class Builder {
+    private Tracer tracer;
+    private int maxAbbreviatedCommandLength = 100;
+
+    /**
+     * Sets the {@link Tracer} to use. Required.
+     */
+    public Builder tracer(final Tracer tracer) {
```

Comment:
move the builder to the top of the file and start like this so as to force the tracer to never be null. Also, we conventionally accept Tracing as that prevents later api breaks when we need other components from there (such as current trace context)

```java
public final class MongoDBTracing {
  public static MongoDBTracing create(Tracing tracing) {
    return newBuilder(tracing).build();
  }

  public static Builder newBuilder(Tracing tracing) {
    return newBuilder(tracing);
  }

  public static Builder newBuilder(Tracing tracing) {
    return new Builder(tracing);
  }

  public static final class Builder {
    final Tracing tracing;
// other fields non-final as mutable
    Builder(Tracing tracing) {
      if (tracing == null) throw new NullPointerException("tracing == null");
      this.tracing = tracing;
--snip--
  
  MongoDBTracing(Builder builder) { // note this is package private
    this.whatINeed = builder...
--snip--

  public CommandListener commandListener() {
    return new MongoDBCommandListener(this); // this type can be package private
  }
```

Reply:
doing like this allows us a possibility to work with multiple versions, which may have multiple types.. In MySQL we use the single type for all because it has to be instantiated statically, which I believe is different here.